bigquery:createWriteStream() should propogate insertErrors

_From @chrishiestand on September 29, 2017 4:30_

The gist is bigquery sometimes silently drops one or more streams when many streams are used in parallel (42 in this example). If I'm hitting a quota an error ought to be thrown, but no errors are thrown the program runs as expected but in the end there are rows missing from the bigquery data.

If there is a bug, it might be in gcloud-node, or it might be in the bigquery api. Both seem less likely than me having made a mistake, so I hope you can find something I've done wrong.

The tricky part is the bug doesn't always reproduce. When the bug does reproduce `n` streams of size `s` are dropped so bigquery will have `n` * `s` missing rows. So if `n=2` and `s=150` bigquery will be missing 300 rows. In other words, the problem does not appear to be that a subset of stream data is missing, but rather one or more entire streams are missing.

This seems to reproduce reliably sometimes, and other times it reliably does not reproduce. To try and get the opposite result, try again later and/or change the stream load with the env variables.

This small bug reproduction project <https://github.com/chrishiestand/gcloud-node-bigquery-manystreams-bug>is the result of troubleshooting missing big query data in a production system where 50 streams are processed in parallel.

Below is a screenshot of the reproduction repository showing a reproduction of the issue.

![image](https://user-images.githubusercontent.com/100825/31000788-2b63b0c2-a493-11e7-9661-ac52fd50b909.png)

In contrast, if I reduce the number of streams from 42 to 10, the tests pass as below:
![image](https://user-images.githubusercontent.com/100825/31000831-925f3134-a493-11e7-89c3-b536445a15a4.png)


#### Environment details

  - OS: OS X 10.12.6
  - Node.js version: 8.6.0
  - npm version: 5.3.0
  - google-cloud-node version: @google-cloud/bigquery = 0.9.6

#### Steps to reproduce

Please go here for a reproduction project: <https://github.com/chrishiestand/gcloud-node-bigquery-manystreams-bug>


_Copied from original issue: GoogleCloudPlatform/google-cloud-node#2635_

_From @chrishiestand on October 2, 2017 19:4_

Thank you very much @stephenplusplus. I've tagged your latest version of the reproduction project as `v0.1.1`.

Now I am finally seeing rate limit errors with `streamConcurrency = 2`, `totalStreams = 100`, and `streamLength = 150`

```js
...
DONE 150
DONE 150
TypeError: Cannot read property 'outputRows' of undefined
    at Job.job.on.job (/private/tmp/gcloud-node-bigquery-manystreams-bug/test/many-streams.test.js:158:37)
    at emitOne (events.js:115:13)
    at Job.emit (events.js:210:7)
    at /private/tmp/gcloud-node-bigquery-manystreams-bug/node_modules/@google-cloud/common/src/operation.js:188:10
    at /private/tmp/gcloud-node-bigquery-manystreams-bug/node_modules/@google-cloud/bigquery/src/job.js:331:5
    at /private/tmp/gcloud-node-bigquery-manystreams-bug/node_modules/@google-cloud/common/src/service-object.js:272:5
    at Object.handleResp (/private/tmp/gcloud-node-bigquery-manystreams-bug/node_modules/@google-cloud/common/src/util.js:135:3)
    at /private/tmp/gcloud-node-bigquery-manystreams-bug/node_modules/@google-cloud/common/src/util.js:465:12
    at Request.onResponse [as _callback] (/private/tmp/gcloud-node-bigquery-manystreams-bug/node_modules/retry-request/index.js:160:7)
    at Request.self.callback (/private/tmp/gcloud-node-bigquery-manystreams-bug/node_modules/request/request.js:186:22)
    at emitTwo (events.js:125:13)
    at Request.emit (events.js:213:7)
    at Request.<anonymous> (/private/tmp/gcloud-node-bigquery-manystreams-bug/node_modules/request/request.js:1163:10)
    at emitOne (events.js:115:13)
    at Request.emit (events.js:210:7)
    at Gunzip.<anonymous> (/private/tmp/gcloud-node-bigquery-manystreams-bug/node_modules/request/request.js:1085:12)
    at Object.onceWrapper (events.js:314:30)
    at emitNone (events.js:110:20)
    at Gunzip.emit (events.js:207:7)
    at endReadableNT (_stream_readable.js:1059:12)
    at _combinedTickCallback (internal/process/next_tick.js:138:11)
    at process._tickCallback (internal/process/next_tick.js:180:9)
{ kind: 'bigquery#job',
  etag: '"Z8nD8CIuj_TqmPrS_MNV-O9x2rU/RoVaRZnuqFcy-MsZbrpdHrL7cy4"',
  id: 'my-project:job_xwxZ4JQ7ESbaCf9Aw0xkFyfVcxl7',
  selfLink: 'https://www.googleapis.com/bigquery/v2/projects/my-project/jobs/job_xwxZ4JQ7ESbaCf9Aw0xkFyfVcxl7',
  jobReference:
   { projectId: 'my-project',
     jobId: 'job_xwxZ4JQ7ESbaCf9Aw0xkFyfVcxl7' },
  configuration:
   { load:
      { destinationTable: [Object],
        sourceFormat: 'NEWLINE_DELIMITED_JSON' } },
  status:
   { state: 'DONE',
     errorResult:
      { reason: 'rateLimitExceeded',
        location: 'table.write',
        message: 'Exceeded rate limits: too many table update operations for this table. For more information, see https://cloud.google.com/bigquery/troubleshooting-errors' },
     errors: [ [Object] ] },
  statistics:
   { creationTime: '1506969989800',
     startTime: '1506969989869',
     endTime: '1506969989869' },
  user_email: 'user@my-project.iam.gserviceaccount.com' }
/private/tmp/gcloud-node-bigquery-manystreams-bug/test/many-streams.test.js:40
  throw error;
  ^

TypeError: Cannot read property 'outputRows' of undefined
    at Job.job.on.job (/private/tmp/gcloud-node-bigquery-manystreams-bug/test/many-streams.test.js:163:39)
    at emitOne (events.js:115:13)
    at Job.emit (events.js:210:7)
    at /private/tmp/gcloud-node-bigquery-manystreams-bug/node_modules/@google-cloud/common/src/operation.js:188:10
    at /private/tmp/gcloud-node-bigquery-manystreams-bug/node_modules/@google-cloud/bigquery/src/job.js:331:5
    at /private/tmp/gcloud-node-bigquery-manystreams-bug/node_modules/@google-cloud/common/src/service-object.js:272:5
    at Object.handleResp (/private/tmp/gcloud-node-bigquery-manystreams-bug/node_modules/@google-cloud/common/src/util.js:135:3)
    at /private/tmp/gcloud-node-bigquery-manystreams-bug/node_modules/@google-cloud/common/src/util.js:465:12
    at Request.onResponse [as _callback] (/private/tmp/gcloud-node-bigquery-manystreams-bug/node_modules/retry-request/index.js:160:7)
    at Request.self.callback (/private/tmp/gcloud-node-bigquery-manystreams-bug/node_modules/request/request.js:186:22)
    at emitTwo (events.js:125:13)
    at Request.emit (events.js:213:7)
    at Request.<anonymous> (/private/tmp/gcloud-node-bigquery-manystreams-bug/node_modules/request/request.js:1163:10)
    at emitOne (events.js:115:13)
    at Request.emit (events.js:210:7)
    at Gunzip.<anonymous> (/private/tmp/gcloud-node-bigquery-manystreams-bug/node_modules/request/request.js:1085:12)
    at Object.onceWrapper (events.js:314:30)
    at emitNone (events.js:110:20)
    at Gunzip.emit (events.js:207:7)
    at endReadableNT (_stream_readable.js:1059:12)
    at _combinedTickCallback (internal/process/next_tick.js:138:11)
    at process._tickCallback (internal/process/next_tick.js:180:9)
  ✖ test/many-streams.test.js exited with a non-zero exit code: 7

  0 tests passed [11:46:30]
  1 uncaught exception

npm ERR! Test failed.  See above for more details.
```

--- a/src/table.ts
+++ b/src/table.ts
@@ -1231,7 +1231,10 @@ Table.prototype.createWriteStream = function(metadata) {
         });
 
         job.metadata = data;
-        dup.emit('complete', job);
+
+        job
+          .on('error', err => dup.destroy(err))
+          .on('complete', () => dup.emit('complete', job));
       }
     );
   });

--- a/src/table.ts
+++ b/src/table.ts
@@ -1212,6 +1212,11 @@ Table.prototype.createWriteStream = function(metadata) {
 
   const dup = streamEvents(duplexify());
 
+  // This will allow the natural `finish` event to fire.
+  dup.on('prefinish', () => {
+    dup.cork();
+  });
+
   dup.once('writing', function() {
     common.util.makeWritableStream(
       dup,
@@ -1245,7 +1250,10 @@ Table.prototype.createWriteStream = function(metadata) {
 
         job
           .on('error', err => dup.destroy(err))
-          .on('complete', () => dup.emit('complete', job));
+          .on('complete', () => {
+            dup.emit('complete', job);
+            dup.uncork();
+          });
       }
     );
   });

--- a/src/table.ts
+++ b/src/table.ts
@@ -1142,9 +1142,12 @@ Table.prototype.createReadStream = paginator.streamify('getRows');
  *
  * request.get(csvUrl)
  *   .pipe(table.createWriteStream(metadata))
- *   .on('complete', function(job) {
+ *   .on('job', function(job) {
  *     // `job` is a Job object that can be used to check the status of the
  *     // request.
+ *   })
+ *   .on('complete', function(job) {
+ *     // The job has completed successfully.
  *   });
  *
  * //-
@@ -1154,7 +1157,13 @@ Table.prototype.createReadStream = paginator.streamify('getRows');
  *
  * fs.createReadStream('./test/testdata/testfile.json')
  *   .pipe(table.createWriteStream('json'))
- *   .on('complete', function(job) {});
+ *   .on('job', function(job) {
+ *     // `job` is a Job object that can be used to check the status of the
+ *     // request.
+ *   })
+ *   .on('complete', function(job) {
+ *     // The job has completed successfully.
+ *   });
  */
 Table.prototype.createWriteStream = function(metadata) {
   const self = this;
@@ -1231,6 +1240,8 @@ Table.prototype.createWriteStream = function(metadata) {
         });
 
         job.metadata = data;
+
+        dup.emit('job', job);
 
         job
           .on('error', err => dup.destroy(err))

--- a/system-test/bigquery.ts
+++ b/system-test/bigquery.ts
@@ -740,10 +740,16 @@ describe('BigQuery', function() {
     });
 
     it('should insert rows via stream', function(done) {
+      let job;
+
       fs.createReadStream(TEST_DATA_JSON_PATH)
         .pipe(table.createWriteStream('json'))
         .on('error', done)
-        .on('complete', function() {
+        .on('complete', function(_job) {
+          job = _job;
+        })
+        .on('finish', function() {
+          assert.strictEqual(job.metadata.status.state, 'DONE');
           done();
         });
     });

--- a/test/table.ts
+++ b/test/table.ts
@@ -1561,7 +1561,7 @@ describe('BigQuery/Table', function() {
           .emit('writing');
       });
 
-      it('should emit complete when job is complete', function(done) {
+      it('should emit finish when the job is complete', function(done) {
         const metadata = {
           jobReference: {
             jobId: 'job-id',
@@ -1571,21 +1571,27 @@ describe('BigQuery/Table', function() {
           c: 'd',
         };
 
-        makeWritableStreamOverride = function(stream, options, callback) {
+        makeWritableStreamOverride = function(dup, options, callback) {
+          dup.setWritable(new stream.PassThrough());
           callback(metadata);
           setImmediate(function() {
             fakeJob.emit('complete');
           });
         };
 
+        let completeWasEmitted = false;
+
         table
           .createWriteStream()
           .on('complete', function(job) {
             assert.strictEqual(job, fakeJob);
-            assert.deepStrictEqual(job.metadata, metadata);
+            completeWasEmitted = true;
+          })
+          .on('finish', function() {
+            assert.strictEqual(completeWasEmitted, true);
             done();
           })
-          .emit('writing');
+          .end('data');
       });
     });
   });

--- a/test/table.ts
+++ b/test/table.ts
@@ -1387,10 +1387,16 @@ describe('BigQuery/Table', function() {
     });
 
     describe('writable stream', function() {
+      let fakeJob;
       let fakeJobId;
 
       beforeEach(function() {
+        fakeJob = new events.EventEmitter();
         fakeJobId = uuid.v4();
+
+        table.bigQuery.job = function(id, options) {
+          return fakeJob;
+        };
 
         fakeUuid.v4 = function() {
           return fakeJobId;
@@ -1407,15 +1413,6 @@ describe('BigQuery/Table', function() {
 
         stream = table.createWriteStream();
         stream.emit('writing');
-      });
-
-      it('should pass the connection', function(done) {
-        makeWritableStreamOverride = function(stream, options) {
-          assert.deepStrictEqual(options.connection, table.connection);
-          done();
-        };
-
-        table.createWriteStream().emit('writing');
       });
 
       it('should pass extended metadata', function(done) {
@@ -1504,17 +1501,22 @@ describe('BigQuery/Table', function() {
         table.createWriteStream(options).emit('writing');
       });
 
-      it('should create a job and emit it with complete', function(done) {
-        const jobId = 'job-id';
+      it('should create a job and emit it with job', function(done) {
         const metadata = {
-          jobReference: {jobId, location: LOCATION},
+          jobReference: {
+            jobId: 'job-id',
+            location: 'location',
+          },
           a: 'b',
           c: 'd',
         };
 
         table.bigQuery.job = function(id, options) {
-          assert.strictEqual(options.location, LOCATION);
-          return {id: id};
+          assert.strictEqual(id, metadata.jobReference.jobId);
+          assert.deepStrictEqual(options, {
+            location: metadata.jobReference.location,
+          });
+          return fakeJob;
         };
 
         makeWritableStreamOverride = function(stream, options, callback) {
@@ -1523,8 +1525,63 @@ describe('BigQuery/Table', function() {
 
         table
           .createWriteStream()
+          .on('job', function(job) {
+            assert.strictEqual(job, fakeJob);
+            assert.deepStrictEqual(job.metadata, metadata);
+            done();
+          })
+          .emit('writing');
+      });
+
+      it('should return an error if the job fails', function(done) {
+        const error = new Error('Error.');
+
+        const metadata = {
+          jobReference: {
+            jobId: 'job-id',
+            location: 'location',
+          },
+          a: 'b',
+          c: 'd',
+        };
+
+        makeWritableStreamOverride = function(stream, options, callback) {
+          callback(metadata);
+          setImmediate(function() {
+            fakeJob.emit('error', error);
+          });
+        };
+
+        table
+          .createWriteStream()
+          .on('error', function(err) {
+            assert.strictEqual(err, error);
+            done();
+          })
+          .emit('writing');
+      });
+
+      it('should emit complete when job is complete', function(done) {
+        const metadata = {
+          jobReference: {
+            jobId: 'job-id',
+            location: 'location',
+          },
+          a: 'b',
+          c: 'd',
+        };
+
+        makeWritableStreamOverride = function(stream, options, callback) {
+          callback(metadata);
+          setImmediate(function() {
+            fakeJob.emit('complete');
+          });
+        };
+
+        table
+          .createWriteStream()
           .on('complete', function(job) {
-            assert.strictEqual(job.id, jobId);
+            assert.strictEqual(job, fakeJob);
             assert.deepStrictEqual(job.metadata, metadata);
             done();
           })

